src/FilenameParsing/FileIDHelperResolutionStrategy.php

```diff
@@ -0,0 +1,458 @@
+    }
+
+    /**
+     * @return FileIDHelper[]
+     */
+    public function getResolutionFileIDHelpers()
+    {
+        return $this->resolutionFileIDHelpers;
+    }
+
+    /**
+     * @param FileIDHelper[] $resolutionFileIDHelpers
+     */
+    public function setResolutionFileIDHelpers(array $resolutionFileIDHelpers)
+    {
+        $this->resolutionFileIDHelpers = $resolutionFileIDHelpers;
+    }
+
+    /**
+     * @return string
+     */
+    public function getVersionedStage()
+    {
+        return $this->versionedStage;
+    }
+
+    /**
+     * @param string $versionedStage
+     */
+    public function setVersionedStage($versionedStage)
+    {
+        $this->versionedStage = $versionedStage;
+    }
+
+
+    public function buildFileID($tuple)
+    {
+        $parsedFileID = $this->preProcessTuple($tuple);
+        return $this->getDefaultFileIDHelper()->buildFileID(
+            $parsedFileID->getFilename(),
+            $parsedFileID->getHash(),
+            $parsedFileID->getVariant()
+        );
+    }
+
+    public function findVariants($tuple, Filesystem $filesystem)
+    {
+        $parsedFileID = $this->preProcessTuple($tuple);
+
+        $helpers = $this->getResolutionFileIDHelpers();
+        array_unshift($helpers, $this->getDefaultFileIDHelper());
+
+        // Search for a helper that will allow us to find a file
+        /** @var FileIDHelper $helper */
+        $helper = null;
+        foreach ($helpers as $helperToTry) {
+            $fileID = $helperToTry->buildFileID(
+                $parsedFileID->getFilename(),
+                $parsedFileID->getHash()
+            );
+            if ($filesystem->has($fileID) && $this->validateHash($helperToTry, $parsedFileID, $filesystem)) {
+                $helper = $helperToTry;
+                break;
+            }
+        }
+
+        if ($helper) {
+
+            $folder = $helper->lookForVariantIn($parsedFileID);
+            $possibleVariants = $filesystem->listContents($folder, true);
+            foreach ($possibleVariants as $possibleVariant) {
+                if ($possibleVariant['type'] !== 'dir' && $helper->isVariantOf($possibleVariant['path'], $parsedFileID)) {
+                    yield $helper->parseFileID($possibleVariant['path'])->setHash($parsedFileID->getHash());
+                }
+            }
+        }
+    }
+    
+    public function cleanFilename($filename)
+    {
+        return $this->getDefaultFileIDHelper()->cleanFilename($filename);
+    }
+    
+    public function parseFileID($fileID)
+    {
+        foreach ($this->resolutionFileIDHelpers as $fileIDHelper) {
+            $parsedFileID = $fileIDHelper->parseFileID($fileID);
+            if ($parsedFileID) {
+                return $parsedFileID;
+            }
+        }
+
+        return null;
+    }
+
+    public function stripVariant($fileID)
+    {
+        $hash = '';
+
+        // Normalise our parameters
+        if ($fileID instanceof ParsedFileID) {
```

Comment:
I added a bit more detail. That's another of those method that received either `FileID`/`ParsedFileID`. We could split it up, but this one makes a little bit more sense, because it's easier to normalise a ParsedFileID to a FileID string.